enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,502 @@
+  filling up. Only once swap is full, the system will cross `memory.high`
+  and trigger soft evictions.
+
+  * Pro
+    * Simple to achieve.
+  * Con
+    * A lot of memory pressure has to be present in order to trigger
+      soft eviction.
+    * Once `memory.high` is reached, the whole `kubepods.slice` is throttled
+      and cannot allocate memory, which might lead to applications crashing.
+
+** **General Availability** - Memory-based soft eviction is going to
+  be disabled, in favor of enabling swap-based hard evictions, based on new
+  swap traffic and swap utilization eviction metrics.
+
+  * Pro
+    * Eviction on the basis of swap pressure, not only memory pressure.
+    * [LLN] applies, because all pods share the nodes memory
+  * Con
+    * Swap-based evictions are made through a 3rd party container, which means
+      it has to be done through an API-initiated eviction.
+
+###### Hypervisor/OS level assistance 
+
+Swap allows us to increase the
+virtual address space. This is required in order to reduce the
+likelihood for workloads to OOM significantly.
+
+Besides _increasing_ the virtual address space, there are two
+mechanisms which help to reduce the memory footprint of a workload.
+
+They are not being discussed in depth here, but are mentioned, as they
+help to reduce or avoid memory pressure, which in turn leads less usage
+of swap.
+
+The mechanisms are:
+* _Free Page Reporting ([FPR])_ - A mechanism to free memory on the
+  hypervisor side if it is not used by the guest anymore.
+* _Kernel Samepage Merging ([KSM])_ - A mechanism to merge/deduplicate
+  memory pages with the same content. This is particularly helpful
+  in situations where similar guests are running on a hypervisor.
+  It has however some security implications.
+
+##### VM higher-density: Differences between Technology Preview vs GA
+
+|                              | TP                  | GA                  |
+|------------------------------|---------------------|---------------------|
+| SWAP Provisioning            | MachineConfig       | MachineConfig       |
+| SWAP Eligibility             | burstable QoS pods  | burstable QoS pods  |
+| Node service protection      | Yes                 | Yes                 |
+| I/O saturation protection    | Yes                 | Yes                 |
+| Critical workload protection | No                  | Yes                 |
+| Memory pressure handling     | Memory based        | Memory & Swap based |
+
+### Risks and Mitigations
+
+#### Phases 1-2
+
+| Risk                                       | Mitigation                                                                                               |
+|--------------------------------------------|----------------------------------------------------------------------------------------------------------|
+| Miss details and introduce instability     | * Adjust overcommit ratio <br/> * Tweak eviction thresholds <br/> * Use de-scheduler to balance the load |
```

Comment:
is anything needed in the descheduler to do this?

Reply:
Added link.